File: sdkwork-im-sdk-flutter/composed/lib/src/streams_module.dart
import 'package:im_sdk_generated/im_sdk_generated.dart';

import 'builders.dart';
import 'context.dart';
import 'types.dart';

class ImStreamsModule {
  final ImSdkContext context;

  ImStreamsModule(this.context);

  Future<StreamSession?> open(OpenStreamRequest body) {
    return context.transportClient.stream.open(body);
  }

  Future<StreamFrameWindow?> listFrames(
    String streamId, [
    ImQueryParams? params,
  ]) {
    return context.transportClient.stream.listStreamFrames(streamId, params);
  }

  Future<StreamFrame?> appendFrame(
    String streamId,
    AppendStreamFrameRequest body,
  ) {
    return context.transportClient.stream.appendStreamFrame(streamId, body);
  }

  Future<StreamFrame?> appendTextFrame(
    String streamId,
    ImAppendTextFrameOptions options,
  ) {
    return appendFrame(
      streamId,
      ImBuilders.textFrame(options),
    );
  }

  Future<StreamSession?> checkpoint(
    String streamId,
    CheckpointStreamRequest body,
  ) {
    return context.transportClient.stream.checkpoint(streamId, body);
  }

  Future<StreamSession?> complete(
    String streamId,
    CompleteStreamRequest body,
  ) {
    return context.transportClient.stream.complete(streamId, body);
  }

  Future<StreamSession?> abort(
    String streamId,
    AbortStreamRequest body,
  ) {
    return context.transportClient.stream.abort(streamId, body);
  }
}

File: sdkwork-im-sdk-rust/composed/src/streams_module.rs
use crate::{build_text_stream_frame, ImSdkContext, QueryParams, TextFrameOptions};
use sdkwork_im_sdk_generated::{
  AbortStreamRequest,
  AppendStreamFrameRequest,
  CheckpointStreamRequest,
  CompleteStreamRequest,
  OpenStreamRequest,
  SdkworkError,
  StreamFrame,
  StreamFrameWindow,
  StreamSession,
};

#[derive(Clone)]
pub struct ImStreamsModule {
  context: ImSdkContext,
}

impl ImStreamsModule {
  pub(crate) fn new(context: ImSdkContext) -> Self {
    Self { context }
  }

  pub async fn open(&self, body: OpenStreamRequest) -> Result<StreamSession, SdkworkError> {
    self.context.transport_client().stream().open(&body).await
  }

  pub async fn list_frames(
    &self,
    stream_id: impl AsRef<str>,
    params: Option<&QueryParams>,
  ) -> Result<StreamFrameWindow, SdkworkError> {
    let stream_id = stream_id.as_ref().to_string();
    self
      .context
      .transport_client()
      .stream()
      .list_stream_frames(&stream_id, params)
      .await
  }

  pub async fn append_frame(
    &self,
    stream_id: impl AsRef<str>,
    body: AppendStreamFrameRequest,
  ) -> Result<StreamFrame, SdkworkError> {
    let stream_id = stream_id.as_ref().to_string();
    self
      .context
      .transport_client()
      .stream()
      .append_stream_frame(&stream_id, &body)
      .await
  }

  pub async fn append_text_frame(
    &self,
    stream_id: impl AsRef<str>,
    frame_seq: i64,
    text: impl Into<String>,
    options: TextFrameOptions,
  ) -> Result<StreamFrame, SdkworkError> {
    self
      .append_frame(stream_id, build_text_stream_frame(frame_seq, text, options))
      .await
  }

  pub async fn checkpoint(
    &self,
    stream_id: impl AsRef<str>,
    body: CheckpointStreamRequest,
  ) -> Result<StreamSession, SdkworkError> {
    let stream_id = stream_id.as_ref().to_string();
    self.context.transport_client().stream().checkpoint(&stream_id, &body).await
  }

  pub async fn complete(
    &self,
    stream_id: impl AsRef<str>,
    body: CompleteStreamRequest,
  ) -> Result<StreamSession, SdkworkError> {
    let stream_id = stream_id.as_ref().to_string();
    self.context.transport_client().stream().complete(&stream_id, &body).await
  }

  pub async fn abort(
    &self,
    stream_id: impl AsRef<str>,
    body: AbortStreamRequest,
  ) -> Result<StreamSession, SdkworkError> {
    let stream_id = stream_id.as_ref().to_string();
    self.context.transport_client().stream().abort(&stream_id, &body).await
  }
}

